Create rule S6893: Use a constant value for the apiVersion (#3778)
SonarSource/rspec: rules/S6893/azureresourcemanager/rule.adoc

The `apiVersion` of a resource is used to determine the configurable properties and values for the resource.
Setting it as a parameter or variable is not recommended, as it can lead to unexpected behaviors and deployment failures.

== Why is this an issue?

> In Azure, different API versions of a resource can have changing properties and values.

Reviewer: ```suggestion
In Azure, different API versions of a resource can have different properties and values.
```

Reviewer 2: I didn't want to have different two times in one sentence, that why I did it this way.

> Using a variable or parameter for the `apiVersion` for a resource may seem like good practice to always stay up to date with

Reviewer: ```suggestion
Using a variable or parameter for the `apiVersion` for a resource is not an optimal way to always stay up to date with
```
My feeling is, we shouldn't say that code smells may seem like a good practice, to keep the tone more neutral.

      "type": "string"
    }
  },
  "resources": [
    {
      "apiVersion": "[parameters('customApiVersion')]",
      "type": "Microsoft.Compute/virtualMachines",
      "name": "nonCompliantResourceWithParameterApiVersion"
    }
  ]
}
----

[source,json]
----
{
  "variables": {
    "customApiVersion": "[first(providers(‘Microsoft.Compute’,’virtualMachines’).apiVersions)]"
  },
  "resources": [
    {
      "apiVersion": "[variables('customApiVersion')]",
      "type": "Microsoft.Compute/virtualMachines",
      "name": "nonCompliantResourceWithVariableApiVersion"
    }
  ]
}
----

==== Compliant solution

[source,json,diff-id=1,diff-type=compliant]
----
{
  "resources": [
    {
      "apiVersion": "2020-06-01",
      "type": "Microsoft.Compute/virtualMachines",
      "name": "compliantResource"
    }
  ]
}
----

== Resources
=== Documentation

* https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/best-practices#api-version[ARM template best practices: API version]
* https://learn.microsoft.com/en-us/azure/templates/[Resource Template Reference]

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

> Use a hardcoded value for the `apiVersion` of this resource.

Reviewer: Maybe, calling it `constant` instead of `hardcoded` could be better? The work "hardcoded" feels like it has a negative connotation to me.

Reviewer 2: I used the same term that is in the ARM best practices...But I agree, I like your suggestion better.
